[ENG-211] Add sidecar support for v1 mempool

wllmshao opened the pull request:
- Create new `ReapedTxs` type which is just a `types.Txs` and a `[]int64 GasWanteds`. This is used as the new return type of the mempool and sidecar reap functions.
- Create a new `mempool/mev_reap.go` file with a function that merges sidecar and mempool `ReapedTxs` up to `maxBytes, maxGas`
- Reset v0 mempool to original tendermint v0 mempool
- Add sidecar channel to v1 mempool reactor
- Create and use sidecar in `node.go` regardless of which mempool version is used

wllmshao committed:
Support v1 mempool

MaghnusM commented on mempool/mev_reap.go:
```diff
@@ -0,0 +1,64 @@
+	sidecarTxsMap := make(map[types.TxKey]struct{})
+
+	for i, sidecarTx := range sidecarTxs.Txs {
+		fmt.Println("[mev-tendermint]: reaped sidecar mev transaction", sidecarTx.Hash())
```
I'd consider shortening, can be heavy real estate on the logs

wllmshao replied:
agreed, switching to `GetLastNumBytesFromTx(sidecarTx, 20)`

wllmshao replied:
Switched to use `getLastNumBytesFromTx`, same for mempltx logging below

MaghnusM commented on mempool/mev_reap.go:
```diff
@@ -0,0 +1,64 @@
+		if maxGas > -1 && newTotalGas > maxGas {
+			return txs
+		}
+		totalGas = newTotalGas
```
May want to consolidate logic here, first case you check with added amount for bytes, then update running total var. Second case you add, then check total, then set running amount to total.

wllmshao replied:
I agree but this is the way original tendermint does it so didn't want to change what isn't broken. Going to leave as is.

MaghnusM commented on mempool/v0/clist_mempool.go:
```diff
@@ -580,13 +554,19 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64, sidecarTxs [
 		if maxGas > -1 && newTotalGas > maxGas {
-			return txs[:len(txs)-1]
+			return types.ReapedTxs{
+				Txs:        txs[:len(txs)-1],
```
isn't this just all of them? just set as txs?

wllmshao replied:
It's exclusive of last index so this will exclude the last element

MaghnusM commented on mempool/v0/clist_mempool.go:
```diff
@@ -580,13 +554,19 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64, sidecarTxs [
-			return txs[:len(txs)-1]
+			return types.ReapedTxs{
+				Txs:        txs[:len(txs)-1],
+				GasWanteds: gasWanteds[:len(txs)-1],
```
same here?

MaghnusM commented on mempool/v1/reactor.go:
```diff
@@ -182,6 +236,81 @@ type PeerState interface {
+				}
+				bz, err := msg.Marshal()
+				if err != nil {
+					panic(err)
```
this will only panic the peer thread and stop it right?

wllmshao replied:
Good question, this will actually panic and there's no recovery on this call stack, so this would take the node down.
I'm removing the panic on both v0 and v1 `broadcastTxSidecarRoutine`s, instead just logging and skipping the send. Don't want a failed sidecar marshal to take down a node.

MaghnusM reviewed (changes requested):
such a good upgrade, well – just some small comments but this is so much cleaner than before

Will look more closely tomorrow

wllmshao committed:
Clean logs, remove broadcast panic

bpiv400 commented on mempool/mev_reap_test.go:
```diff
@@ -0,0 +1,34 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+func TestCombineSidecarAndMempoolTxs_SkipsSidecarTxsInMempoolReap(t *testing.T) {
```
I think we should also add tests to ensure that if the side car exceeds max gas and/or max bytes, things work as expected. 

Same for the case where the mempool plus the sidecar exceeds those values but mempool alone does not

wllmshao replied:
Good point, added these tests.

bpiv400 commented on mempool/v0/clist_mempool.go:
```diff
@@ -691,3 +671,19 @@ func (mem *CListMempool) recheckTxs() {
 //--------------------------------------------------------------------------------
+
+// mempoolTx is a transaction that successfully ran
+type mempoolTx struct {
```
Why did we move this out of mempool.go and into here + why did we lowercase the fields?

wllmshao replied:
I fully reset this to original tendermint, where `mempoolTx` is private and only part of v0.
We made a public `MempoolTx` for our sidecar implementation for mev-tendermint, which is no longer necessary.

bpiv400 commented on mempool/v1/reactor.go:
```diff
@@ -155,25 +166,68 @@ func (memR *Reactor) RemovePeer(peer p2p.Peer, reason interface{}) {
 // It adds any received transactions to the mempool.
 func (memR *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
```
Not looking at this super closely -- assuming we're pretty much copy pasting the logic here

wllmshao replied:
Yeah, we're copying. But an additional change piggybacked on this is that if the `MEVMessage` marshal fails in `broadcastSidecarTxRoutine`, we no longer panic, just log and continue. This is changed in both v0 and v1 reactors.

bpiv400 reviewed (changes requested):
Overall, very good IMO.

Just some structural things: 
1. Why did we change the mempoolTx data structure names and location from mempool.go to v0 mempool? It seems like this belongs in the types along with ReapedTxs package even though we only use it in v0 now, it looks like. 
2. If sidecar is used for both v0 and v1, it probably shouldn't go into the v0 package. It seems like it belongs in its own package? Seems strange that v1 now imports the v0 package in a couple locations. 

And a note about testing: It seems like the combined reap logic is undertested right now, since we don't have tests around the edge and failure cases related to max bytes and max gas being exceeded in the sidecar or in the mempool after the sidecar.

wllmshao committed:
Move clist_sidecar to mempl and add reactor v1 unit tests

wllmshao committed:
Add mev_reap unit tests

wllmshao commented:
> Why did we change the mempoolTx data structure names and location from mempool.go to v0 mempool? It seems like this belongs in the types along with ReapedTxs package even though we only use it in v0 now, it looks like.

See comment reply above.

> If sidecar is used for both v0 and v1, it probably shouldn't go into the v0 package. It seems like it belongs in its own package? Seems strange that v1 now imports the v0 package in a couple locations.

Agreed with this, moved the sidecar into mempool, out of v0. Also removed its dependence on the v0 `mempoolTx`

> And a note about testing: It seems like the combined reap logic is undertested right now, since we don't have tests around the edge and failure cases related to max bytes and max gas being exceeded in the sidecar or in the mempool after the sidecar.

Agreed, added more tests for mev_reap. Also added the sidecar tests from v0 reactor to v1 reactor.
